Templates

GlebFedorov opened the pull request:
# Added jinja2 template support. 
Docs now can be built by filling out templates with data from yaml files.

To add new template you simply need to create a jinja2 template in docs/reach/templates/ with mdx extension and add an entry to templates dictionary in one or several reach*.yml config files in following format.
```yaml
templates:  # <- this should already be in the file
    destination: 'template_name.mdx'
```
Jinja2 template design docs: https://jinja.palletsprojects.com/en/2.10.x/templates/ 

# Rendering

Templates can be rendered via python script:
```bash
pipenv install 
cd script
pipenv run python build.py <tatget>
# Where <target> is .yml config file.
# example
pipenv run python build.py ../reachm-plus.yml 
```
By default script updates doc files in given models's "docs_dir" directory, which can be found in reach*.yml config files.

GlebFedorov committed:
templates: create reach power supply template

GlebFedorov committed:
config_rtk.yml: create build config for reach rtk

GlebFedorov committed:
config_mplus.yml: create build config for reach mplus

GlebFedorov committed:
build.py: add template building script

build.py takes a path to .yml file as an argument and prints
generated text to stdout based on a template from templates/ directory

GlebFedorov committed:
conf: move config files to separate directory

GlebFedorov committed:
tempaltes: reach_power_rs.md: add template for reach rs/rs2 docs

GlebFedorov committed:
conf: config_mplus.yml add index template support

GlebFedorov committed:
templates: index_template.md: add index template

GlebFedorov committed:
conf: config_mplus.yml: fix indentation

GlebFedorov committed:
conf: config_rs2.yml: add index template support

GlebFedorov committed:
config_mplus.yml: config_rs2.yml: index_template.md: move image markdown syntax from config to the template

GlebFedorov committed:
conf: config_rs.yml: config_mplus.yml: add RS/RS+ config file

GlebFedorov committed:
conf: config_rs.yml: update model name

GlebFedorov committed:
templates: index_template.md: fix model name templating

GlebFedorov committed:
templates: index_template.md: index_rtk_template.md: add RTK specific template and include it

templates:  add RTK template include

GlebFedorov committed:
conf: config_rtk.yml: add RTK config

GlebFedorov committed:
update_index.py: add template building script

Script updates index.md file for every provided config

GlebFedorov committed:
templates: index_rtk_template.md: fix template text

GlebFedorov committed:
templates: reach_power.md: add RTK include

GlebFedorov committed:
render.py: add renderer class for building doc from specified configs or templates

GlebFedorov committed:
templates: add include directory, move templates called only as includes to said directory

GlebFedorov committed:
conf: rs2.yml: rs.yml: add usb type data

GlebFedorov committed:
script: add script directory and move template building scripts to said directory

GlebFedorov committed:
README.md: add basic template rendering info

GlebFedorov committed:
docs: reach: templates: introduce new file extension specific to markdown templates

GlebFedorov committed:
Pipfile: add template building dependencies

GlebFedorov committed:
script: build.py: add template builder script

script takes path to yaml file as an argument and updates all templates associated with given file

GlebFedorov committed:
README.md: add template building info

GlebFedorov committed:
docs: reach: templates: clean templates directory, apply bettter naming convention

GlebFedorov committed:
reach.yml: reachm-plus.yml: reachrs.yml: reachrs2.yml: add templates entry

templates: is a dictionarry containing template building data
where destination file is key and template name is value.

GlebFedorov committed:
script: build.py: rework template-destination pairing

Building script now simply parses dictionarry provided in reach*.yml files

staroselskii commented on reachm-plus.yml:
```diff
@@ -83,3 +83,23 @@ nav:
+  - 'img/reachm-plus/antenna-placement.png'
+  - 'img/reachm-plus/power-supply.png'
+  - 'img/reachm-plus/specs.png'
+inthebox:
```
Maybe it should be `kit`?

GlebFedorov committed:
dreachm-plus.yml: reachrs.yml: reachrs2.yml: docs: reach: templates: index.mdx: rename inthebox to kit

 rename inthebox to kit

GlebFedorov committed:
docs: templates: move templates to docs directory

GlebFedorov committed:
READMT.md: add template workflow info

staroselskii reviewed (changes requested):
Please, rebase and squash commits that don't add any value.

GlebFedorov committed:
templates: move templates directory to docs/reach

GlebFedorov committed:
reach.yml: reachm-plus.yml: reachrs.yml: reachrs2.yml: move config info to main yaml files, delete redundant configs

GlebFedorov committed:
docs: reach: templates: index.mdx: power-supply.mdx: remove unnecessary includes

GlebFedorov committed:
build.py: move build.py to the root directory

staroselskii commented:
@TatianaAndreeva please take a look.

TatianaAndreeva commented on README.md:
````diff
@@ -9,16 +9,51 @@ We use mkdocs to build the docs and Pipenv to manage the dependencies.
+```yaml
+templates:  # <- this should already be in the file
+    destination: '<template-name.mdx>'
+```
````
Can we add an example here?

GlebFedorov replied:
Added contents of reachm-plus.yml as an example.

TatianaAndreeva replied:
I think it's better to format that in such way:

```yaml
templates:  # <- this should already be in the file
    destination: '<template-name.mdx>'
```

Where the destination is the name of the file in **docs_dir** (docs_dir is specified in each .yml file) directory, that is going to be rendered via the template.

And template_name name is the name of the actual template file with .mdx extension that must be placed inside **emlid-docs/doc/templates** directory.

**Without these entries template will be ignored.**

Example:

```yaml
templates:
    index.md: "index.mdx"
    power-supply.md: "power-supply.mdx"
```

TatianaAndreeva replied:
Could you please distinguish the example as I suggested above? I think it makes this easier to understand for the person who opens this file for the first time

TatianaAndreeva commented on README.md:
````diff
@@ -9,16 +9,51 @@ We use mkdocs to build the docs and Pipenv to manage the dependencies.
+# example
+pipenv run python build.py reachm-plus.yml 
+```
+## Adding new templates
````
I believe it's better to swap "Adding new templates" and "Updating docs after editing templates (Generating markdown)" parts

GlebFedorov replied:
Makes sense.

TatianaAndreeva reviewed (changes requested):
Why do we have 2 similar files here:
- docs/templates/index.mdx,
- docs/reach/templates/index_template.md?

Also, I'd like to point out that we don’t use “_” in file naming. Instead of it, please name the files with “-”.

GlebFedorov committed:
doc: reach: templates: remove redundant templates

GlebFedorov committed:
README.md: add an example of template to doc pairing

GlebFedorov committed:
README.md: swap "Adding new templates" and "Update docs after editing templates"

TatianaAndreeva commented on README.md:
```diff
@@ -9,16 +9,50 @@ We use mkdocs to build the docs and Pipenv to manage the dependencies.
+
+## Updating docs after editing templates (Generating markdown)
+
+Template building script takes path to .yml file as an argument and rebuilds docs from every template specified in .yml.
```
Template building script takes path to .yml file as an argument and rebuilds docs from every template specified in .yml. => Template building script takes the path to .yml file as an argument and rebuilds docs from every template specified in .yml.

TatianaAndreeva commented on reach.yml:
```diff
@@ -80,3 +80,20 @@ nav:
+        - 'Micro-USB <-> USB cable'
+        - 'Micro-USB <-> USB-F OTG cable'
+templates:
+         power-supply.md: 'power-supply.mdx'
```
Why do we specify the template name here in single quotes? In the README file, double quotes are used.

TatianaAndreeva commented on reachm-plus.yml:
```diff
@@ -83,3 +83,23 @@ nav:
+  - 'img/reachm-plus/power-supply.png'
+  - 'img/reachm-plus/specs.png'
+kit:
+  - "* 1 x Reach M+ module"
```
Here are double quotes as well. I believe we should use the same format everywhere

GlebFedorov replied:
Fixed including previously existing inconsistencies.

staroselskii commented on reachm-plus.yml:
```diff
@@ -83,3 +83,23 @@ nav:
+model: M+
+power_port: JST-GH
+index_images:
+  - 'img/reachm-plus/quickstart.png'
```
`index_images` shouldn't be here as they have the same name (or can be made into having the same name). You can just use `img/${model}/tutorials.png` img/${model}/reachview.png` et al. in the `index.mdx`

GlebFedorov replied:
Added new variable `model_path` since changing `model` would replace model name in actual docs and renaming directory with images to "M+" is a  bad idea.

GlebFedorov committed:
README.md: edge.yml: navio.yml: navio2.yml: reach.yml: reachm-plus.yml: reachrs.yml: reachrs2.yml:

fix quotes to be consistent with the rest of the configs

GlebFedorov committed:
README.md: fix grammar

GlebFedorov committed:
docs: reach: rs: img: reachrs: Placement.png: Power-supply.png: Quickstart.png: ReachView.png : Specs.png: Tutorials.png:

docs: reach: rs2: img: reachrs2: Placement.png: Power-supply.png: Quickstart.png: ReachView.png : Specs.png: Tutorials.png:

rename images to lower case letters to be consistent with m+

GlebFedorov committed:
reachm_plus.yml: reachrs.yml: reachrs2.yml: replace index_images with model_path

GlebFedorov committed:
docs: templates: index.mdx: replace for loop with explicit entries with variables

GlebFedorov committed:
README.md: separate format and example in "Adding new templates"

giragleb commented:
I do the following steps:

1. Make a .mdx file in the templates folder 
2. Add it in reachrs2.yml
3. Update docs by `pipenv run python build.py reachrs2.yml`
4. Run locally by `pipenv run mkdocs serve -f reachrs2.yml`

Have the following warnings: 

WARNING -  Config value: 'power_port'. Warning: Unrecognised configuration name: power_port 
WARNING -  Config value: 'kit'. Warning: Unrecognised configuration name: kit 
WARNING -  Config value: 'model'. Warning: Unrecognised configuration name: model 
WARNING -  Config value: 'templates'. Warning: Unrecognised configuration name: templates 
WARNING -  Config value: 'model_path'. Warning: Unrecognised configuration name: model_path

So I can't see any changes in the docs. 

Should I do something to recognize templates?

GlebFedorov commented:
@giragleb 
> WARNING - Config value: 'power_port'. Warning: Unrecognised configuration name: power_port
> WARNING - Config value: 'kit'. Warning: Unrecognised configuration name: kit
> WARNING - Config value: 'model'. Warning: Unrecognised configuration name: model
> WARNING - Config value: 'templates'. Warning: Unrecognised configuration name: templates
> WARNING - Config value: 'model_path'. Warning: Unrecognised configuration name: model_path
> 
> 
> Should I do something to recognize templates?

No, you shouldn't. Mkdocs simply complains about unused entries.  We are storing both templates and mkdocs configurations in one file, so these warnings indicate that there are no conflicts in two configs. 

> So I can't see any changes in the docs.

If any changes have been made, you should see them in markdown file right after running the build.py script, so you don't really need to build html pages to roughly estimate if templates are rendered correctly.

Just to be sure, I did try to create a new template and everything seems to work fine. So, please make sure that:
1. Your .mdx template is in **/docs/templates/** folder
2.  That your destination markdown file is in **docs_dir**  (for reach RS2 it's /docs/reach/rs2) if it is not, it will be created with rendered content in that directory.
3.  That you are running build.py script from **emlid-docs** directory.

If this doesn't fix the problem, I'll require some more details to be able help you.

giragleb commented:
@GlebFedorov,

I tried to reproduce this from scratch, and everything works fine now. 

The most incomprehensible thing is, that I should do `run python build.py reach*.yml` every time I want to see any changes locally. 

For example, if I do `run python build.py reachrs.yml` and then `pipenv run mkdocs serve -f reachrs2.yml`, I'll get docs for reachrs in the *.md file. That means that I should do 2 steps to change between docs.

Can we simplify or automate this process somehow?

GlebFedorov commented:
@giragleb,

Sorry for the late replies, school has been pretty liberal with my time this week.

> That means that I should do 2 steps to change between docs.
> 
> Can we simplify or automate this process somehow?
 
In other words you want to be able to automatically serve page right after building markdown?
If that's so, yes, it can be added as an option.

>The most incomprehensible thing is, that I should do ```run python build.py reach*.yml``` every time I want to see any changes locally.

I am afraid, that you will have to run the script each time to render the templates.

Sorry, if I am pointing out the obvious,  but these commands can be made less tedious by activating virtual environment  by running ```pipenv shell```.  While it's active you can leave ```pipenv run``` part out of all commands. (```exit``` to deactivate it)

giragleb commented:
> In other words you want to be able to automatically serve page right after building markdown? If that's so, yes, it can be added as an option.

Well, I meant that we want to automatically build markdown before serving the page, but the essence seems to be the same. If you can add this option, that would be great.

GlebFedorov committed:
update_serve.py: add script that serves static pages right after rendering templates

giragleb commented:
This seems to work fine with your examples, however, I can't make it work with my created template.

I can make my example work using `python build.py reach*.yml` and `mkdocs serve -f reach*.yml`, but this doesn't work using only _serve_ with a new script.

I believe that I'm missing some step, can you please update README.md according to the new feature?

giragleb commented:
Hi @GlebFedorov 

Do you need any input from me in regards to the previous question?

GlebFedorov committed:
README.md: add update_serve.py info

GlebFedorov commented:
@giragleb,

I somehow managed to completely miss your response. I am really sorry.

I added new script to the README.

As for your problem, it's really strange. This script doesn't really do anything but executing both commands one by one. Are you running it via `pipenv`? It works fine for me.

giragleb commented:
It works now with `pipenv run python update_serve.py <target>.yml`, thanks.
